File: src/21.ts
import _ from 'lodash';
import sigma from './core/sigma';

// Computes the sum of all amicable numbers under `limit`.
//
// See {@link https://projecteuler.net/problem=21}
export default function compute(limit: number) {
  let sum = 0;

  for (const a of _.range(2, limit + 1)) {
    const b = sigma(a) - a;
    if (a > b && sigma(b) - b === a) {
      sum += a + b;
    }
  }

  return sum;
}

File: project_euler/21.py
# Problem 21: Amicable Numbers
#
# Evaluate the sum of all the amicable numbers under 10_000.
#
# See https://projecteuler.net/problem=21
from project_euler.common.divisors import sigma


def run(n: int) -> int:
    result = 0

    for a in range(2, n + 1):
        b = sigma(a) - a
        if a > b and sigma(b) - b == a:
            result += a + b

    return result


def test_run():
    assert run(10_000) == 31626